eclair-core/src/main/scala/fr/acinq/eclair/channel/fund/InteractiveTxBuilder.scala

```diff
@@ -247,9 +251,9 @@ object InteractiveTxBuilder {
     case class Remote(serialId: UInt64, amount: Satoshi, pubkeyScript: ByteVector) extends Output with Incoming
 
     /** The shared output can be added by us or by our peer, depending on who initiated the protocol. */
-    case class Shared(serialId: UInt64, pubkeyScript: ByteVector, localAmount: MilliSatoshi, remoteAmount: MilliSatoshi) extends Output with Incoming with Outgoing {
+    case class Shared(serialId: UInt64, pubkeyScript: ByteVector, localAmount: MilliSatoshi, remoteAmount: MilliSatoshi, htlcsAmount: MilliSatoshi) extends Output with Incoming with Outgoing {
```

Comment:
This seems a bit surprising that we need this here, and that we don't split into local/remote HTLCs...are you sure this is necessary? If so, how come we don't need the same thing for the shared input as well?

I'm wondering if it wouldn't be simpler to include HTLCs inside the `localAmount` and `remoteAmount` (for example by considering that outgoing HTLCs are added to the `remoteAmount` and incoming HTLCs are added to the `localAmount`, as if we're expecting them to be fulfilled), and only when we create the commitment txs at the very end, we recreate the HTLCs by deducing them from these `localAmount` and `remoteAmount` by using the `purpose` field.

I'm not entirely sure, but I think it's worth trying that approach to avoid over-complicating the `Input` and `Output` classes with details that we could keep internal.

Reply:
I've refactored out he `htlcsAmount` parameter from Output.Shared in [504317d](https://github.com/ACINQ/eclair/pull/2720/commits/504317d2a7542509851ba0d12ddaa40464ee9da1). 

The way I was looking at the `htlcsAmount` was that it was the value that must exist in the new funding tx after local and remote balance changes have been applied. If splice out takes the balance below the amount needed to satisfy the in-flight htlcs then one of the parties could be unable to collect. I also now explicitly check that at the point we make sure the funding output is not below the dust limit.

Reply:
After looking into it more, I think storing the `htlcAmount` actually makes a lot of sense, I tried that in https://github.com/ACINQ/eclair/commit/5b7fe763945829b57a1a4ff9e8d0aafdd5075f3b

I still want to add a bit more tests and review the existing ones (in another commit), but we'll be able to compare which approach is simplest once you've finished your investigation.

Reply:
I have added more commits to my branch (https://github.com/ACINQ/eclair/tree/splice-with-htlcs) to improve the tests. If including the `htlcAmount` is the right direction forward, you should work on top of my test changes, which highlights that force-close tests weren't interesting because HTLCs were trimmed: we need to verify that we correctly claim the on-chain HTLC txs.

I've also grouped tests together, which makes it easier to see the distinction between the tests that have HTLCs and the pre-quiescence tests. I think the tests themselves can be improved by being more consistent.

Reply:
In [05ef07d](https://github.com/ACINQ/eclair/pull/2720/commits/05ef07df8177c5e5ae6572f8131c35ad95068b94) I show how we can refactor `Input.Shared` `Output.Shared` to unify local/remote/shared amounts (millisatoshi) into a single amount (satoshi).  

This change primarily impacted:
 - computing `localFees` (used for tests and reporting command success during dual funding)
 - logic of `shouldSignFirst` (I also noticed that this already deviates from initial draft splice spec)
 - `InteractiveTxBuilderSpec` needs further review to make sure local/remote fees are properly checked
 - I am not certain local/remote pushes of value are properly accounted for everywhere; needs more tests

This investigation shows that combining the values for local/remote/shared amounts has minimal downside. After this change we only store these amounts in one place (pre/post splice commitments) and allow the shared inputs/outputs to use whole satoshi amounts.

If we decide to go with this approach, I'll need to integrate the test changes from @t-bast , and generally focus on more tests.